pkg/acpi/coreboot/madt.go

```diff
@@ -0,0 +1,169 @@
+	}
+	b = b[8:]
+	Debug("NEWMADT Subtables: after removing 8 bytes of fluff, %d bytes", len(b))
+	m := &MADT{Table: t, base: t.Address()}
+	for len(b) > 0 {
+		Debug("NewMADT loop: %d bytes\n", len(b))
+		if len(b) < 2 {
+			return nil, fmt.Errorf("subtable is too short: have %d bytes, need 2", len(b))
+		}
+		l := b[1]
+		typ := madtType(b[0])
+		Debug("subtable: %d bytes, l is %#x, t is %#x", len(b), l, t)
+		if int(typ) > len(info) || typ == entryBOGUS {
+			Debug("Bad type: %#x, offset %#x, skipping %d bytes", typ, len(t.TableData())-len(b), l)
+			b = b[l:]
+			continue
+		}
+		if l != info[typ].size {
+			return nil, fmt.Errorf("Type %s has size %d should be %d", info[typ].name, l, info[typ].size)
+		}
+		if len(b) < int(l) {
+			return nil, fmt.Errorf("Type %s has buf length %d should be %d", info[typ].name, len(b), info[typ].size)
+		}
+		m.subtables = append(m.subtables, subtable{t: typ, b: b})
+		b = b[l:]
+	}
+	Debug("NEWMADT Subtables: %d bytes tab %d bytes", len(b), len(m.TableData()))
+	return m, nil
+}
+
+// CoreBoot generates the coreboot C code for a MADT.
+// MADT consists of TLV entries: 1-byte type, 1-byte length, value.
+func (m *MADT) Coreboot(w io.Writer) error {
+	s := fmt.Sprintf(`unsigned long acpi_fill_madt(unsigned long current)
+{
+	/* create all subtables for processors */
+`)
+	// OK this is a bit nasty. Right now coreboot is trying to fill out
+	// the lapic address and flags. So we ignore those in the table for now.
+	// If it's getting them wrong, we will have to go back and
+	// fill them in.
+	//lapic(%#x, %#x);\n", b[:4], b[4:8])
+	// The coreboot serialization code for ACPI uses the C compiler to drive serialization.
+	// It uses this pattern:
+	// int acpi_create_madt_lapic(acpi_madt_lapic_t *lapic, u8 cpu, u8 apic);
+	// The first arg is a uintptr cast to a struct pointer of the table type.
+	// The function stores the arguments into the packed struct, the idea being that
+	// the compiler generates code to store into the struct that exactly matches
+	// the ACPI standard.
+	for _, tab := range m.subtables {
+		b := tab.b
+		switch tab.t {
+		case entryLAPIC:
+			// TODO: Check b[4] to see if the LAPIC is enabled? When would it not be, if it's in the table?
+			s += fmt.Sprintf("\tcurrent += acpi_create_madt_lapic((acpi_madt_lapic_t *)current, %#x, %#x);\n", b[2], b[3])
+		case entryIOAPIC:
+			s += fmt.Sprintf("\tcurrent += acpi_create_madt_ioapic((acpi_madt_ioapic_t *)current, %#x, %#x, %#x);\n", b[2], rv(b[4:8]), rv(b[8:12]))
+		case entryISOR:
+			s += fmt.Sprintf("\tcurrent += acpi_create_madt_irqoverride((acpi_madt_irqoverride_t *)current, %#x, %#x, %#x, %#x);\n", b[2], b[3], rv(b[4:8]), rv(b[8:10]))
+		case entryNMI:
+			s += fmt.Sprintf("\tcurrent += acpi_create_madt_lapic_nmi((acpi_madt_lapic_nmi_t *)current, %#x, %#x, %#x);\n", b[2], rv(b[3:5]), b[5])
+		case entryLAPICOverRide:
+			// coreboot appears sot to support this.
```

Comment:
s/sot/not

happens a few times in this function

Reply:
coding late at night. Oops.